TiFlash late materialization (#12939)
pingcap/docs: tiflash/tiflash-late-materialization.md

---
title: TiFlash Late Materialization
summary: Describe how to use the TiFlash late materialization feature to accelerate queries in OLAP scenarios.
---

# TiFlash Late Materialization

> **Warning:**
>
> Currently, this is an experimental feature. The form and usage might be modified in future releases.

This document describes how to use the TiFlash late materialization feature to accelerate queries in OLAP scenarios.

> By default, TiFlash reads all the data from the columns required by a query, and then filters and aggregates the data based on the query conditions. Late materialization is an optimization method that supports pushing down part of the filtering conditions to the TableScan operator. That is, TiFlash first scans the column data related to one of the filtering conditions, filters out the rows that meet the condition, and then scans the other column data of these rows for further calculation, thus reducing the amount of scan I/O and data parsing computation.

Reviewer: ```suggestion
By default, TiFlash reads all the data from the columns required by a query, and then filters and aggregates the data based on the query conditions. Late materialization is an optimization method that supports pushing down part of the filtering conditions to the TableScan operator. That is, TiFlash first scans the column data related to the pushed down filtering conditions, finds the rows that meet the condition, and then scans the other column data of these rows for further calculation, thus reducing the amount of scan I/O and data processing computation.
```

`filters out` 是过滤掉，不准确。

> By default, when receiving a query request, TiFlash reads all the data from the columns required by the query, and then filters and aggregates the data based on the query conditions. Late materialization is an optimization method that supports pushing down part of the filtering conditions to the TableScan operator. That is, TiFlash first scans the column data related to the filtering conditions that are pushed down, filters the rows that meet the condition, and then scans the other column data of these rows for further calculation, thereby reducing IO scans and computations of data processing.

Reviewer 2: ```suggestion
By default, when receiving a query request, TiFlash reads all the data from the columns required by the query, and then filters and aggregates the data based on the query conditions. Late materialization is an optimization method that supports pushing down part of the filtering conditions to the TableScan operator. That is, TiFlash first scans the column data related to the filtering conditions that are pushed down, filters the rows that meet the condition, and then scans the other column data of these rows for further calculation, thereby reducing computations of IO scans and data processing.
```

> If you want to improve the performance of certain queries in OLAP scenarios, you can enable the TiFlash late materialization feature at the session level or global level. By modifying the value of the [`tidb_opt_enable_late_materialization`](/system-variables.md#tidb_opt_enable_late_materialization-new-in-v700) system variable, you can choose to enable or disable the TiFlash delayed materialization feature.

Author: ```suggestion
If you want to improve the performance of certain queries in OLAP scenarios, you can enable the TiFlash late materialization feature at the session level or global level. By modifying the value of the [`tidb_opt_enable_late_materialization`](/system-variables.md#tidb_opt_enable_late_materialization-new-in-v700) system variable, you can choose to enable or disable the TiFlash late materialization feature.
```

> After enabling the TiFlash late materialization feature, the TiDB optimizer will determine which filtering condition will be pushed down based on statistics and the query's filtering conditions. The optimizer will prioritize pushing down the filtering conditions with high filtering rates. For detailed algorithms, see the [RFC document](https://github.com/pingcap/tidb/tree/master/docs/design/2022-12-06-support-late-materialization.md).

Reviewer: ```suggestion
After enabling the TiFlash late materialization feature, the TiDB optimizer will determine which filtering conditions will be pushed down based on statistics. The optimizer will prioritize pushing down the filtering conditions with low selectivity. For detailed algorithms, please check the [RFC document](https://github.com/pingcap/tidb/tree/master/docs/design/2022-12-06-support-late-materialization.md).
```

Author: "see" 是“参考”文档的通用用法，英文用户文档避免使用 please

> 

Author: "下推过滤率高"，英文对应 “low selectivity”？

Reviewer: 高过滤率对应“low selectivity”

Reviewer: 或者 “high filtration rate”, 不是 “filtering rate”

For example:

```sql
EXPLAIN SELECT a, b, c FROM t1 WHERE a < 1;
```

```
+-------------------------+----------+--------------+---------------+-------------------------------------------------------+
| id                      | estRows  | task         | access object | operator info                                         |
+-------------------------+----------+--------------+---------------+-------------------------------------------------------+
| TableReader_12          | 12288.00 | root         |               | MppVersion: 1, data:ExchangeSender_11                 |
| └─ExchangeSender_11     | 12288.00 | mpp[tiflash] |               | ExchangeType: PassThrough                             |
|   └─TableFullScan_9     | 12288.00 | mpp[tiflash] | table:t1      | pushed down filter:lt(test.t1.a, 1), keep order:false |
+-------------------------+----------+--------------+---------------+-------------------------------------------------------+
```

> In this example, the `a < 1` filtering condition is pushed down to the TableScan operator. TiFlash first reads all data from column `a`, and then filters the rows that meet the `a < 1` condition. Next, TiFlash reads the data of columns `b` and `c` for these rows.

Reviewer: ```suggestion
In this example, the filtering condition `a < 1` is pushed down to the TableScan operator. TiFlash first reads all data from column `a`, and then finds the rows that meet the `a < 1` condition. Next, TiFlash reads the specific rows of columns `b` and `c`.
```

```

```sql
SHOW GLOBAL VARIABLES LIKE 'tidb_opt_enable_late_materialization';
```

```
+--------------------------------------+-------+
| Variable_name                        | Value |
+--------------------------------------+-------+
| tidb_opt_enable_late_materialization | OFF   |
+--------------------------------------+-------+
```

You can modify the `tidb_opt_enable_late_materialization` variable at the session level or at the global level.

- To enable TiFlash late materialization in the current session, use the following statement:

    ```sql
    set session tidb_opt_enable_late_materialization=ON;
    ```

- To enable TiFlash late materialization at the global level, use the following statement:

    ```sql
    set global tidb_opt_enable_late_materialization=ON;
    ```

    After this setting, the `tidb_opt_enable_late_materialization` variable will be enabled by default for both session and global levels in new sessions.

To disable TiFlash late materialization, use the following statements:

```sql
set session tidb_opt_enable_late_materialization=OFF;
```

```sql
set global tidb_opt_enable_late_materialization=OFF;
```

## Implementation mechanism

> When a filtering condition is pushed down to the TableScan operator, the execution process of the TableScan operator mainly includes the following steps:

Reviewer: ```suggestion
When there are some filtering conditions are pushed down to the TableScan operator, the execution process of the TableScan operator mainly includes the following steps:
```

1. Reads the three columns `<handle, del_mark, version>`, performs multi-version concurrency control (MVCC) filtering, and then generates the MVCC Bitmap.

> 2. Reads the columns related to the filtering condition, performs the filtering condition, and then generates the Filter Bitmap.

Reviewer: ```suggestion
2. Reads the columns related to the filtering conditions, performs the filtering conditions, and then generates the Filter Bitmap.
```

3. Performs an `AND` operation between the MVCC Bitmap and Filter Bitmap to generate the Final Bitmap.
4. Reads the corresponding rows of the remaining columns according to the Final Bitmap.

> 5. Returns the results.

Reviewer 2: 漏译了一句话。中文原文：
5. 合并两次读取的数据，返回结果。

Author: added in https://github.com/pingcap/docs/pull/12939/commits/3e749f9cdf4788fea291c514fc5f1bade8ac9092

> 5. Returns the results.

Author: ```suggestion
5. Merges the data read in steps 2 and 4, and then returns the results.
```
